package/src/nvidiamon.cpp

```diff
@@ -0,0 +1,207 @@
+          if (!activegpus.count(gpu_idx)) {
+            ++nvidia_stats["ngpus"];
+            activegpus[gpu_idx] = true;
+          }
+        }
+      }
+    } else if (NVIDIA_DEBUG) {
+      std::clog << "Bad read of line: " << s << std::endl;
+      std::cout << "Parsed to: " << gpu_idx << " " << pid << " " << cg_type
+                << " " << sm << " " << mem << " " << enc << " " << dec << " "
+                << fb_mem << " " << cmd_name << std::endl;
+      std::cout << "StringStream status: good()=" << instr.good();
+      std::cout << " eof()=" << instr.eof();
+      std::cout << " fail()=" << instr.fail();
+      std::cout << " bad()=" << instr.bad() << std::endl;
+    }
+  }
+
+  if (NVIDIA_DEBUG) {
+    std::cout << "Parsed: " << nvidia_stats["ngpus"] << " "
+              << nvidia_stats["gpusmpct"] << " " << nvidia_stats["gpumempct"]
+              << " " << nvidia_stats["gpufbmem"] << std::endl;
+  }
+
+  // Update the statistics with the new snapshot values
+  ++iterations;
+  for (const auto& nvidia_param : nvidia_params) {
+    if (nvidia_stats[nvidia_param] > nvidia_peak_stats[nvidia_param]) {
+      nvidia_peak_stats[nvidia_param] = nvidia_stats[nvidia_param];
+    }
+    nvidia_total_stats[nvidia_param] += nvidia_stats[nvidia_param];
+    nvidia_average_stats[nvidia_param] =
+        double(nvidia_total_stats[nvidia_param]) / iterations;
+  }
+}
+
+// Return the nvidiaers
+std::map<std::string, unsigned long long> const nvidiamon::get_text_stats() {
+  return nvidia_stats;
+}
+
+// For JSON return the peaks
+std::map<std::string, unsigned long long> const
+nvidiamon::get_json_total_stats() {
+  return nvidia_peak_stats;
+}
+
+// An the averages
+std::map<std::string, double> const nvidiamon::get_json_average_stats(
+    unsigned long long elapsed_clock_ticks) {
+  return nvidia_average_stats;
+}
+
+bool nvidiamon::test_nvidia_smi() {
+  const std::vector<std::string> cmd = {"nvidia-smi", "-L"};
+
+  // The use of execvp means searching along PATH, which is fine
+  // but does imply some extra stat() calls; nvidia-smi isn't going
+  // to change location, so there is an argument for finding and
+  // caching the directory path
+  auto cmd_result = prmon::cmd_pipe_output(cmd);
+  if (cmd_result.first != 0) return false;
+  unsigned int gpus = 0;
+  for (auto const& s : cmd_result.second) {
+    // From C++20 can use 'starts_with'
+    if (s.substr(0, 3).compare("GPU") == 0) {
+      ++gpus;
+    }
+  }
+  nvidiamon::ngpus = gpus;
+  if (gpus == 0) {
+    std::clog << "Executed 'nvidia-smi -L', but no GPUs found" << std::endl;
+    return false;
+  } else if (gpus > 4) {
+    std::clog << "More than 4 GPUs found, so GPU process monitoring will be "
+                 "unreliable"
+              << std::endl;
+  }
+  return true;
+}
+
+// Collect related hardware information
+void const nvidiamon::get_hardware_info(nlohmann::json& hw_json) {
+  // Record the number of GPUs
+  hw_json["HW"]["gpu"]["nGPU"] = nvidiamon::ngpus;
```

Comment:
Getting this w/ a constant `get` method would be nicer.

Reply:
Do you think it's really worth it for a private member variable?